# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/debugger/Frame.cpp
# Commit: 2d470b425ce9
# Full Hash: 2d470b425ce9d8fc12c89f087d76ff1a337619ab
# Author: Logan Smyth <loganfsmyth@gmail.com>
# Date: 2020-06-15 21:48:38
# Description:
#   Bug 1644699 - Allow Debugger.Frame to have GeneratorInfo that is closed. r=arai
#   
#   The GeneratorInfo data associated with a Debugger.Frame is not guaranteed
#   to be cleared when a generator transitions from running/suspended to closed.
#   In the current codebase, we have broadly assumed that to be true, which was
# ==============================================================================

diff -r d56e40dc9c0f -r 2d470b425ce9 js/src/debugger/Frame.cpp
--- a/js/src/debugger/Frame.cpp	Mon Jun 15 14:33:28 2020 +0000
+++ b/js/src/debugger/Frame.cpp	Mon Jun 15 14:26:35 2020 +0000
@@ -255,7 +255,7 @@
   }
 
   if (maybeGenerator) {
-    if (!frame->setGenerator(cx, maybeGenerator)) {
+    if (!frame->setGeneratorInfo(cx, maybeGenerator)) {
       return nullptr;
     }
   }
@@ -333,6 +333,11 @@
   }
 };
 
+bool js::DebuggerFrame::isSuspended() const {
+  return hasGeneratorInfo() &&
+         generatorInfo()->unwrappedGenerator().isSuspended();
+}
+
 js::AbstractGeneratorObject& js::DebuggerFrame::unwrappedGenerator() const {
   return generatorInfo()->unwrappedGenerator();
 }
@@ -343,8 +348,8 @@
 }
 #endif
 
-bool DebuggerFrame::setGenerator(JSContext* cx,
-                                 Handle<AbstractGeneratorObject*> genObj) {
+bool DebuggerFrame::setGeneratorInfo(JSContext* cx,
+                                     Handle<AbstractGeneratorObject*> genObj) {
   cx->check(this);
 
   Debugger::GeneratorWeakMap::AddPtr p =
@@ -403,8 +408,8 @@
   return true;
 }
 
-void DebuggerFrame::clearGenerator(JSFreeOp* fop) {
-  if (!hasGenerator()) {
+void DebuggerFrame::clearGeneratorInfo(JSFreeOp* fop) {
+  if (!hasGeneratorInfo()) {
     return;
   }
 
@@ -428,10 +433,10 @@
   fop->delete_(this, info, MemoryUse::DebuggerFrameGeneratorInfo);
 }
 
-void DebuggerFrame::clearGenerator(
+void DebuggerFrame::clearGeneratorInfo(
     JSFreeOp* fop, Debugger* owner,
     Debugger::GeneratorWeakMap::Enum* maybeGeneratorFramesEnum) {
-  if (!hasGenerator()) {
+  if (!hasGeneratorInfo()) {
     return;
   }
 
@@ -444,7 +449,7 @@
     owner->generatorFrames.remove(&info->unwrappedGenerator());
   }
 
-  clearGenerator(fop);
+  clearGeneratorInfo(fop);
 }
 
 /* static */
@@ -457,7 +462,7 @@
       callee = referent.callee();
     }
   } else {
-    MOZ_ASSERT(frame->hasGenerator());
+    MOZ_ASSERT(frame->isSuspended());
 
     callee = &frame->generatorInfo()->unwrappedGenerator().callee();
   }
@@ -477,7 +482,7 @@
 
     result = iter.isFunctionFrame() && iter.isConstructing();
   } else {
-    MOZ_ASSERT(frame->hasGenerator());
+    MOZ_ASSERT(frame->isSuspended());
 
     // Generators and async functions can't be constructed.
     result = false;
@@ -551,7 +556,7 @@
       env = GetDebugEnvironmentForFrame(cx, iter.abstractFramePtr(), iter.pc());
     }
   } else {
-    MOZ_ASSERT(frame->hasGenerator());
+    MOZ_ASSERT(frame->isSuspended());
 
     AbstractGeneratorObject& genObj =
         frame->generatorInfo()->unwrappedGenerator();
@@ -591,7 +596,7 @@
       result = script->pcToOffset(pc);
     }
   } else {
-    MOZ_ASSERT(frame->hasGenerator());
+    MOZ_ASSERT(frame->isSuspended());
 
     AbstractGeneratorObject& genObj =
         frame->generatorInfo()->unwrappedGenerator();
@@ -639,7 +644,7 @@
       }
     }
   } else {
-    MOZ_ASSERT(frame->hasGenerator());
+    MOZ_ASSERT(frame->isSuspended());
 
     // If the frame is suspended, there is no older frame.
   }
@@ -651,7 +656,11 @@
 /* static */
 bool DebuggerFrame::getAsyncPromise(JSContext* cx, HandleDebuggerFrame frame,
                                     MutableHandleDebuggerObject result) {
-  if (!frame->hasGenerator()) {
+  MOZ_ASSERT(frame->isOnStack() || frame->isSuspended());
+
+  if (!frame->hasGeneratorInfo()) {
+    // An on-stack frame may not have an associated generator yet when the
+    // frame is initially entered.
     result.set(nullptr);
     return true;
   }
@@ -702,7 +711,7 @@
       }
     }
   } else {
-    MOZ_ASSERT(frame->hasGenerator());
+    MOZ_ASSERT(frame->isSuspended());
 
     AbstractGeneratorObject& genObj =
         frame->generatorInfo()->unwrappedGenerator();
@@ -744,7 +753,7 @@
       return DebuggerFrameType::WasmCall;
     }
   } else {
-    MOZ_ASSERT(frame->hasGenerator());
+    MOZ_ASSERT(frame->isSuspended());
 
     return DebuggerFrameType::Call;
   }
@@ -799,7 +808,7 @@
       frame->maybeDecrementStepperCounter(cx->runtime()->defaultFreeOp(),
                                           referent);
     }
-  } else if (frame->hasGenerator()) {
+  } else if (frame->isSuspended()) {
     RootedScript script(cx, frame->generatorInfo()->generatorScript());
 
     if (handler) {
@@ -1241,7 +1250,7 @@
 
   // Connections between dying Debugger.Frames and their
   // AbstractGeneratorObjects should have been broken in DebugAPI::sweepAll.
-  MOZ_ASSERT(!frameobj.hasGenerator());
+  MOZ_ASSERT(!frameobj.hasGeneratorInfo());
 
   frameobj.freeFrameIterData(fop);
   OnStepHandler* onStepHandler = frameobj.onStepHandler();
@@ -1264,7 +1273,7 @@
     onPopHandler->trace(trc);
   }
 
-  if (hasGenerator()) {
+  if (hasGeneratorInfo()) {
     generatorInfo()->trace(trc, *this);
   }
 }
@@ -1367,7 +1376,7 @@
 }
 static bool EnsureOnStackOrSuspended(JSContext* cx, HandleDebuggerFrame frame) {
   MOZ_ASSERT(frame);
-  if (!frame->isOnStack() && !frame->hasGenerator()) {
+  if (!frame->isOnStack() && !frame->isSuspended()) {
     JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
                               JSMSG_DEBUG_NOT_ON_STACK_OR_SUSPENDED,
                               "Debugger.Frame");
@@ -1514,7 +1523,7 @@
       script = framePtr.script();
     }
   } else {
-    MOZ_ASSERT(frame->hasGenerator());
+    MOZ_ASSERT(frame->isSuspended());
     script = frame->generatorInfo()->generatorScript();
   }
   // The async promise value is only provided for async functions and
@@ -1591,7 +1600,7 @@
       }
     }
   } else {
-    MOZ_ASSERT(frame->hasGenerator());
+    MOZ_ASSERT(frame->isSuspended());
   }
 
   result.set(nullptr);
@@ -1767,7 +1776,7 @@
       scriptObject = debug->wrapScript(cx, script);
     }
   } else {
-    MOZ_ASSERT(frame->hasGenerator());
+    MOZ_ASSERT(frame->isSuspended());
     RootedScript script(cx, frame->generatorInfo()->generatorScript());
     scriptObject = debug->wrapScript(cx, script);
   }
@@ -1805,7 +1814,7 @@
 }
 
 bool DebuggerFrame::CallData::terminatedGetter() {
-  args.rval().setBoolean(!frame->isOnStack() && !frame->hasGenerator());
+  args.rval().setBoolean(!frame->isOnStack() && !frame->isSuspended());
   return true;
 }
 